### DISCUSSION TITLE
关于 Progress 组件开发相关

### DISCUSSION THREAD

**Discussion by Tyh2001** [FightingDesign/fighting-design#64] [General]
# Progress 进度条

## Attributes

- `percentage` 百分比
- `type` 类型
- `square` 方形的
- `linear` 线性的
- `show-text` 是否展示进度百分比
- `color` 自定义颜色
- `background` 背景色
- `width` 自定义宽度
- `height` 自定义高度

## Event

- `full` 当到达满进度的回调

## 设计稿

https://www.figma.com/file/Z0iGT78cxmEiPsZhZN75C2/Progress-%E8%BF%9B%E5%BA%A6%E6%9D%A1?node-id=0%3A1

## 参考

可以通过下面进行参考设计组件

https://developer.mozilla.org/zh-CN/docs/Web/HTML/Element/Input](https://github.com/FightingDesign/fighting-design/blob/master/start/src/Progress.vue)

**Comment by lzyaom** [comment]
type 是指设计稿中的状态吗？
是的话，感觉改成 status 会好一点

**Reply by Tyh2001** [reply]
这个就是展示不同的类型，就像按钮组件一样，统一都是使用 type 的

**Reply by lzyaom** [reply]
好的

**Comment by lzyaom** [comment]
插槽
- default -- 用于自定义展示内容 ，参数为 { percentage }

**Comment by ding139725** [comment]
有个小建议就是现有的type改为status  然后线形、方形、等用type进行控制，否则线形和方形都可以进行设置，会不会出现不可控制的情况

**Reply by Tyh2001** [reply]
因为之前所有组件都使用了 type 进行实现不同的样式。

线性和方形的话，使用 state 实现是合理的，也可以实现，但是大多数组件其实我是想避免通过字符串来实现，因为字符串是不可控的，如果传入字符串是不合法的，就会产生不可预期的效果，布尔类型相对应更好控制，我的理解哈～

**Comment by Tyh2001** [comment]
[#68](https://github.com/FightingDesign/fighting-design/pull/68) 一些小问题

- `props` 默认值的返回值类型应该是小写的，`ts` 大多数类型都是小写的写法

```ts
// 不合法的
const Props = {
  type: {
    type: String,
    default: (): String => 'primary',
  },
}

// 合法的
const Props = {
  type: {
    type: String,
    default: (): string => 'primary',
  },
}
```

- 可选参数需要加入校验

```ts
import type { PropType } from 'vue'

type buttonType = 'default' | 'primary' | 'success' | 'danger' | 'warning'

const Props = {
  type: {
    type: String as PropType<buttonType>,
    default: (): buttonType => 'default',
    validator: (val: buttonType): boolean => {
      return (
        ['default', 'primary', 'success', 'danger', 'warning'] as const
      ).includes(val)
    },
  },
}
```

其它没问题，很优秀

**Reply by ding139725** [reply]
respect

**Comment by Tyh2001** [comment]
另外在模板中就可以直接使用 `prop` 的任何属性了，不需要再 `prop.xxx` 了

**Reply by ding139725** [reply]
respect

**Comment by lzyaom** [comment]
百分比没有垂直居中，要修改一下样式

**Reply by lzyaom** [reply]
原来的：
<img width="190" alt="image" src="https://user-images.githubusercontent.com/26430638/179363216-eba939c0-3ed5-46c2-9000-4f7c1880d084.png">
现在的：
<img width="196" alt="image" src="https://user-images.githubusercontent.com/26430638/179363273-5c5ac3b8-1bab-462e-b059-a285c7795590.png">

**Reply by Tyh2001** [reply]
好滴，我看下

**Reply by lzyaom** [reply]
我改了，我提交一下吧

**Reply by Tyh2001** [reply]
ok

**Reply by Tyh2001** [reply]
测试一下哈~

还有个自定义高度的配置项

```html
<f-progress type="primary" :percentage="50" :height="40" />
<f-progress type="success" :percentage="50" />
```

**Reply by lzyaom** [reply]
👌

**Reply by lzyaom** [reply]
height 在 [15, 18) 的时候，显示会有点问题
不然添加一个属性，控制是在右侧显示还是在内侧显示百分比？

**Reply by Tyh2001** [reply]
可以的 新增一个配置项吧
@lzyaom
